File: src/Web.Blazor/Services/Calendar/MonthGrid.cs
namespace SafetyScale.Web.Blazor.Services.Calendar;

/// <summary>Calendar grid helpers. Parity with React <c>monthGrid.ts</c>.</summary>
public static class MonthGrid
{
    public static readonly string[] WeekdayLabelsSmtwtfs = ["S", "M", "T", "W", "T", "F", "S"];

    public static string TodayKeyLocal()
    {
        var today = DateTime.Now;
        return DateKeyFromLocal(today.Year, today.Month, today.Day);
    }

    public static string DateKeyFromLocal(int year, int month, int day) =>
        $"{year:D4}-{month:D2}-{day:D2}";

    public static IReadOnlyList<MonthGridCell> BuildMonthGrid(int viewYear, int viewMonth0)
    {
        var first = new DateTime(viewYear, viewMonth0 + 1, 1);
        var startPad = (int)first.DayOfWeek;
        var daysInMonth = DateTime.DaysInMonth(viewYear, viewMonth0 + 1);
        var cells = new List<MonthGridCell>();

        var prevLast = DateTime.DaysInMonth(
            viewMonth0 == 0 ? viewYear - 1 : viewYear,
            viewMonth0 == 0 ? 12 : viewMonth0);
        for (var i = 0; i < startPad; i++)
        {
            var dayNum = prevLast - startPad + i + 1;
            var prevMonth0 = viewMonth0 == 0 ? 11 : viewMonth0 - 1;
            var prevYear = viewMonth0 == 0 ? viewYear - 1 : viewYear;
            var d = new DateTime(prevYear, prevMonth0 + 1, dayNum);
            cells.Add(new MonthGridCell(DateKeyFromLocal(d.Year, d.Month, d.Day), dayNum, false));
        }

        for (var day = 1; day <= daysInMonth; day++)
        {
            cells.Add(new MonthGridCell(
                DateKeyFromLocal(viewYear, viewMonth0 + 1, day),
                day,
                true));
        }

        var rem = cells.Count % 7;
        var tail = rem == 0 ? 0 : 7 - rem;
        var nextMonth0 = viewMonth0 == 11 ? 0 : viewMonth0 + 1;
        var nextYear = viewMonth0 == 11 ? viewYear + 1 : viewYear;
        for (var n = 1; n <= tail; n++)
        {
            var d = new DateTime(nextYear, nextMonth0 + 1, n);
            cells.Add(new MonthGridCell(DateKeyFromLocal(d.Year, d.Month, d.Day), n, false));
        }

        return cells;
    }
}

public sealed record MonthGridCell(string Key, int Label, bool InMonth);

File: archive/legacy-react-web/src/shared/components/calendar/monthGrid.ts
/** Local date key yyyy-MM-dd (no timezone drift for calendar grids). */

function pad2(n: number): string {
  return String(n).padStart(2, '0')
}

export function dateKeyFromLocal(d: Date): string {
  return `${d.getFullYear()}-${pad2(d.getMonth() + 1)}-${pad2(d.getDate())}`
}

export function todayKeyLocal(): string {
  return dateKeyFromLocal(new Date())
}

export type MonthGridCell = {
  key: string
  label: number
  inMonth: boolean
}

export function buildMonthGrid(viewYear: number, viewMonth0: number): MonthGridCell[] {
  const first = new Date(viewYear, viewMonth0, 1)
  const startPad = first.getDay()
  const daysInMonth = new Date(viewYear, viewMonth0 + 1, 0).getDate()
  const cells: MonthGridCell[] = []

  const prevLast = new Date(viewYear, viewMonth0, 0).getDate()
  for (let i = 0; i < startPad; i++) {
    const dayNum = prevLast - startPad + i + 1
    const d = new Date(viewYear, viewMonth0 - 1, dayNum)
    cells.push({ key: dateKeyFromLocal(d), label: dayNum, inMonth: false })
  }

  for (let day = 1; day <= daysInMonth; day++) {
    const d = new Date(viewYear, viewMonth0, day)
    cells.push({ key: dateKeyFromLocal(d), label: day, inMonth: true })
  }

  const rem = cells.length % 7
  const tail = rem === 0 ? 0 : 7 - rem
  let n = 1
  const ny = viewMonth0 === 11 ? viewYear + 1 : viewYear
  const nm = viewMonth0 === 11 ? 0 : viewMonth0 + 1
  for (let i = 0; i < tail; i++) {
    const d = new Date(ny, nm, n)
    cells.push({ key: dateKeyFromLocal(d), label: n, inMonth: false })
    n++
  }

  return cells
}

export const WEEKDAY_LABELS_SMTWTFS = ['S', 'M', 'T', 'W', 'T', 'F', 'S'] as const
